Restructure operator to allow easy support for  multi-container pod deployment

wyzhang opened the pull request:
Re-structuring operator to introduce "internal" crd (sort of), so CR gets mapped to "internal" CR which  then used to fill the template and generate stateful/deployment/services. 

spec.go
- This is where "internal" CRD is defined 

Note that 
- Manual tested. UI loads. 
- Upgrade/downgrade logic is missing. Yet to be added.
- Some fields in template haven't been set or set correctly yet (e.g. env). Will fix in follow-up changes.

wyzhang committed:
Refacoring: move from one controller compoment per service model to all services in a single component.
This change is in preparation for introduce "numPods" option in CRD which allows users to specify how many pod/deployment/statefulset they want CDAP services to be distributed across

wyzhang committed:
Refactoring: remove Objects and Observables for individual services
as we have moved to single controller-reconciler component model
that all services get reconciled in a single handle.

Note setNodePort() gets updated to match observed with expected
by both namespace and name, otherwise it might find obseved user interface
service and try to set the port on expected router service.

wyzhang committed:
Adding ContainerSpec{} StatelessSpec{} StatefulSpec{} structs
(Not being used yet. Implementation is incomplete and code is untested)

These structs will be used as the internal representation of CRD
i.e. take CR => convert to a list of StatelessSpec{} and StatefulSpec{}
 => used to fill the yaml template to produce expected resource state.

wyzhang committed:
Deleting all handlers and revert back to the state of just having a skeleton.

Will re-write each handler based on the new structure in follow-up commits.

Fixing a few issues in spec.go so that the code compiles and runs. But operator
does nothing in this commit.

wyzhang committed:
Adding back reconciling handler for ConfigMap. Manual test passes.

Major changes:
- Added "ConfigMapSpec" that captures all info for building a configmap obj
- Created "type ConfigMap struct{}" as reconciling handler for ConfigMap.

Minor changes:
- Rename "ServiceType" to "ServiceName" as AppFabric/Messaging etc is more
like a service name as opposed to a type of service. Also avoid confusion
with k8s ServiceType (e.g. NodePort)

wyzhang committed:
Adding back reconciling for most stateful and stateless services
(e.g logs, messaging, metrics, preview, appfab, metadata).
Manul test passed.

wyzhang committed:
Adding back "router" pod and service. Manual test passed.

wyzhang committed:
Adding back all services including UI. Manual test passed (UI load)

wyzhang committed:
Code cleanup and refactoring: a bunch of renamings and code cleanups

barney-s commented on templates/cdap-ui-deployment.yaml:
```diff
@@ -14,55 +14,52 @@
-    {{range $k,$v := .Labels}}
-    {{$k}}: '{{$v}}'
+    {{range $k,$v := .Base.Labels}}
+    {{$k}}: {{$v}}
```
dont you need the single quotes ?

wyzhang replied:
Not sure. Some of our existing template use single quotes. Others don't. Seem to work without them, thus removing to make it consistent.

From quick search
https://stackoverflow.com/questions/19109912/do-i-need-quotes-for-strings-in-yaml
looks to be quotes are in generally not required

barney-s commented on templates/cdap-ui-deployment.yaml:
```diff
@@ -14,55 +14,52 @@
       labels:
-        {{range $k,$v := .Labels }}
-        {{$k}}: '{{$v}}'
+        {{range $k,$v := .Base.Labels }}
```
same single quote comment

barney-s commented on templates/cdap-ui-deployment.yaml:
```diff
@@ -14,55 +14,52 @@
-        - image: {{.Master.Status.UserInterfaceImageToUse}}
+        - name: {{$c.Name}}
+          image: {{$c.Image}}
           workingDir: "/opt/cdap/ui"
```
is working dir same for all containers (non ui ones)

wyzhang replied:
Yes. For this specific template, it is used by our UI service, which will be a normal single-container pod.

barney-s reviewed (commented):
first pass

wyzhang committed:
Code refactoring: moving constants and ApplyDefault from v1alpha1 to controllers.

Specially
- constants.go:
  constants was declared in v1alpha1 package but used by controllers package, thus moving
- utils.go:
  move some util functions to this file
- Fixes
  type ServiceName string is missing "=" before string, otherwise it becomes a typedef
  (we need is a type alias)
- Removed cdapmaster.go
  ApplyDefault() there should be together with controller, thus being moved.

wyzhang committed:
Adding support for configurable number of Pods. Manual test passed.
// 0/unset:  Each service runs in its own Pod
// 1: All services run in a single multi-container Pod
// 2: UI in its own Pod. Other services in a single multi-container Pod
// 3: UI and Router in their own Pod. All other services runs in a multi-container Pod

Other notable change:
- Use deployment template for UI and get rid of UI specific template

Minor changes
- Renaming funcitons and vars.
- Code cleaning up
- Splitting code into multiple files

wyzhang committed:
Adding support for NumPods == 3. Missed in previous commit. Manual test passed.

wyzhang committed:
Re-write image verion upgrade and downgrade logic

wyzhang committed:
Cleanup and minor fixes for the last commit of re-writing upgrading and downgrading logic

wyzhang committed:
Fixing build breakage. Manual tested. Upgrading and downgrade passed.

wyzhang committed:
Set the propagationPolicy in controller-reconciler lib, so that it deletes the child pod when deleting job.
Without this change, pre-/post- upgrade job would leave orphan Pod after job gets deleted

wyzhang committed:
Adding tests for version_update.go

Minor changes
- Create symlink to template dir in controller/ so tests can call functions that
  expect template files to be under <working dir>/template
- Remov ttl in upgrade-job.yaml so we can use jsondiff's superset match
  Also there is no need for TTL as reconciler logic should take care of deleting
  jobs once it either succeeds or fails.

wyzhang committed:
Adding unit tests for CDAP deployment strategy numPods == 0

wyzhang committed:
Refactoring: Convert all tests to use Ginkgo framework

wyzhang committed:
Adding back the setting of java max heap size env var in yaml files

Also add tests and updating exsiting tests.

wyzhang committed:
Adding logic to extract ServiceAccountName, RuntimeClassName and PriorityClassName
from a list of CDAP service spec. This is needed to support multi-container pod
where all services to be placed in the same pod should have the same settings
for these fields, otherwise returns an error if the values are set inconsistently
across services in the same pod.

wyzhang committed:
Code refactoring and cleanup.
- A bunch of renamings
- Extracting some code into functions
- Adding comments

wyzhang committed:
Fixing a TODO about setting the right value for StorageClass and StorageSize.
- Introduce two util functions to extract StorageClass across a list of services
- Calculate the aggregated storage size across a list of services
- Small code refactoring to change function signature

wyzhang committed:
Code cleanup: re-ordering functions to group related ones together

wyzhang committed:
Code cleaning: remove util functions for parsing byte size strings.
Have already switched to k8s resource.Quantify

wyzhang commented:
Made a bunch of major revisions. The change getting much bigger. 
Major changes are: 
* Re-wrote the upgrade/downgrade logic 
* Added unit tests. Decent test coverage. More to be added. 
* Broke code into multiple files

With all these changes, it achieves the same functionality as the current "develop" branch + configurable # of pods deployment

wyzhang commented:
One more major changes forgot to mention:
- Removed the UI yaml template. Made the deployment template flexible that can be adopted by UI

wyzhang committed:
Code refactoring: using reflection to get CDAPSpecService field

wyzhang committed:
Fixing a typo in the description of a test case

wyzhang committed:
Code refactoring: using reflect to get CDAPStatefulServiceSpec and CDAPExternalServiceSpec
- Fixing TODOs
- Adding tests

wyzhang commented:
Some code refactoring: using reflect to get pointer to struct field, thus simplify the code

wyzhang committed:
Code cleanup: some minor coding style cleanup

wyzhang committed:
Adding PreUpgradeFailed and PostUpgradeFailed conditions to track pre-/post- upgrade job failures.

jayfjin commented on controllers/version_update.go:
```diff
@@ -0,0 +1,489 @@
+	switch compareVersion(curVersion, newVersion) {
+	case -1:
+		// Upgrade:
+		updateStatus.clearAllConditions(master)
```
Is this correct? Won't this clear the status each reconcile loop, thus defeating the purpose of tracking the state.

wyzhang replied:
Updated the code based on offline discussion. 

At the high level, make cdap operator not retrying upgrade. So if upgrade fails, leave all conditions in their state for user to exam. 

So the logic to check the upgrade status and react to it is:
if (upgrade inflight) { wait }
if (upgrade failed condition is set) {
  if (spec.image == status.image) {
     return upgrade_failed
  } else {
    return bad_state (because image upgraded but post-upgrade processing failed)
  }
}
return upgrade_succeeded.

jayfjin replied:
lgtm

wyzhang committed:
Change from retry version update to no retry, so leave it in the state can be examed.

Specially,
When upgraded failed, don't reset all conditions and don't retry upgrading.
To retry upgrade, user needs to reset the image version in spec, so this triggers
condition resetting. Then user upgrades the image version again which trigger a fresh new upgrade.

jayfjin reviewed (approved):
lgtm for upgrade based on our discussions

wyzhang committed:
Adding comment to explain the state of conditions and image for upgrade/downgrade success or failure cases
